### PR TITLE
Enable write_files and scripts_user modules

### PR DESCRIPTION
**Pull request by codello**
In this pull request I propose to enable the modules `write_files` and `scripts_user` by default. I also propose to move the `runcmd` module into the `cloud_config_modules` stage. This addresses two (probably quite common) use cases.

One possible use case of `vyos-cloud-init` is to supply a custom VyOS configuration via user data. With the changes in this pull request the defaults would allow users to supply their own configuration file via the `write_files` module (see also #11).

Another possible use case is to supply a custom `vbash` script to configure the system at boot time. Previously the `runcmd` module was enabled for this case. The module however does not actually run the supplied commands unless the `scripts_user` module is enabled as well. For consistency with the `cloud-init` defaults I suggest moving the `runcmd` module into the `cloud_config_modules` section.

### CODE REVIEW

**Comment by codello** [comment]
Hey @mitechie, I think that this PR sets some sensible default values for the `vyos-cloud-init` package. Maybe someone has the time to take a quick look at it.

**Comment by zdc** [comment]
Hello, @codello!
Thanks for this PR!

Preconfiguration via User-Data is something that we have planned to add during this year. Almost one year ago we prepared the User-Data handler for this purpose, but the testing stage delayed a lot till that moment.
Today I have ported the most valuable part of this handler to the main package, so (I hope) that the most wanted feature - the ability to add a full configuration is done now.

Please, see details here: https://github.com/vyos/vyos-cloud-init/pull/27
I hope that now we will be able to get rid of the `runcmd` and `scripts_user` modules.

If you see any non-covered by the new module features, please inform me so we can extend it or make them work in another way.

**Comment by codello** [comment]
Hey @zdc, thanks for your response. I haven't tested the updated code yet but from a quick glance this seems to solve the problem of updating the VyOS config via commands. I think it's a bit unfortunate that this implementation is kind of VyOS specific (relying on a separate key in the cloud-config) but I realize that running a `vbash`  script in the early boot stages is not that easy (I have only been successful using the workaround described in https://github.com/vyos/vyos-cloud-init/issues/11#issuecomment-738333081). Even though this approach is a little less flexible (I don't think you can use variables and control statements as you could in a `vbash` script) I think it is sufficient for most cases.

However I still think the second part of this PR (enabling the `write_files` module) would be useful. At least for me being able to supply a whole configuration file is one of the main uses of cloud-init for VyOS (that way I can very easily create a VyOS instance in Terraform). I think it would be a useful default setting to have `write_files` enabled. Would you like me to amend this PR accordingly?

Alternatively I could look into writing a simple part handler that would enable processing `#vyos-config` (or similar) parts directly. I don't have much experience with cloud-init but I think something like this should be possible and would solve this usecase in a very elegant way. Using a part handler is probably already possible by suppling a multipart user-data document. But maybe it would be appropriate to include such a part handler in this package by default. Let me know what you think.

**Comment by zdc** [comment]
Thanks for the details, @codello!

In VyOS 1.3 (the `current` branch of this repo), we try to be as close to the upstream and common Cloud-init behavior, as is possible. And provide some VyOS extras on top of this.
Currently, implemented already:
Standardized [Meta-Data v1 format](https://cloudinit.readthedocs.io/en/latest/topics/instancedata.html#format-of-instance-data-json) processing - it is used to configure hostname and credentials.
[Network-config](https://cloudinit.readthedocs.io/en/latest/topics/network-config.html) - Ethernets type only. I have not found any platform that uses other types (VLAN, Bonding), so have not added them yet. Via it can be configured all the interfaces, routes, DNS servers, and MTU.
For most cases, this is enough - it is possible to get a router with credentials and all interfaces/ routes configured. And this all config should be received from a virtual environment transparently for users.

For customers who want to configure the router completely, we would like to allow using User-Data to pass a complete config. And, again, we would not like to reinvent anything here that would not be already familiar to Cloud-init users. Therefore, we simply added a new module for processing standard User-Data format. User-Data itself can be provided in any way that is possible in Cloud-init. It is also possible to use [Jinja templates](https://cloudinit.readthedocs.io/en/latest/topics/instancedata.html#using-instance-data).
This is not very obvious, but using commands instead of a complete config file is safer for customers. But it still easy - there is no big difference between:
`cat /config/config.boot`
and
`show configuration commands` (op-mode)
`show | commands` (config-mode)
Even more, we prefer to use commands-style to exchange configuration in our work, because it generates a shorter output that is easy to modify and can be applied to the router via simple Ctrl+C -> Ctrl+V.
Of course, this does not mean that providing a config file is a bad idea. It only requires more attention and care without visible benefits.

If you know any use cases, where this is not enough, can you show them to me? I know at least one, that will be described below, but maybe you know something more?

About write_files. I think it can be useful because allows passing files like keys for OpenVPN, certificates for IPSec, etc. And this is definitely something that is not possible via the config commands only.
So, currently, I agree that including `write files` into `cloud_config_modules` is a good idea.

About custom part-handler. This is a dead-end, I have [passed it already](https://github.com/zdc/vyos-cloud-init-userdata/).
It is good for testing purposes, one-time tasks, or if you cannot modify the main Cloud-init package.
But it is pretty easy to get an even better level of comfort with VyOS-specific User-Data entries. There simply no visible drawbacks here. They are available for modules without modification of the upstream Cloud-init sources, so we keeping compatibility and get all the features that Cloud-init provides automatically.

Summarizing:

- if you agree with the rest, can you update the PR to include only write_files so we can merge it?
- I would be very appreciative if you provide a detailed example of a case when any custom scripts are still required after including `vyos_config_commands`. The situation here is the next: most of them might be solved by easier ways and we would be happy to provide these ways. So, the more detailed use cases will be - the more features appear.

**Comment by codello** [comment]
Sorry for the late reply. Thank you very much for your detailed response. This has cleared up some confusion for me.
> This is not very obvious, but using commands instead of a complete config file is safer for customers. But it still easy - there is no big difference between:

Ah this is something I didn't realize. Thank you. I was under the impression that providing a config file was actually the preferred way. In this case your approach makes a lot of sense to me.

> About custom part-handler. This is a dead-end, I have passed it already.

I see. I agree that the (admittedly small) convenience of providing a `#vyos-config` document does not justify the added complexity of a non-standard part handler.

> I would be very appreciative if you provide a detailed example of a case when any custom scripts are still required after including vyos_config_commands. The situation here is the next: most of them might be solved by easier ways and we would be happy to provide these ways. So, the more detailed use cases will be - the more features appear.

I think `vyos_config_commands` solves a lot of common use cases. One use case for custom scripts might be situations in which the scripts already exist and would need to be converted into the `vyos_config_commands` format. I don't think this is a strong enough argument to justify the added complexity and for me personally this is not relevant.

The only actually important use case I can think of is getting secrets into the VyOS config. I'm not yet very familiar with ways to manage secrets with cloud-init but I think it's bad practice to include sensitive data as clear-text in the `user-data` of a cloud provider. One approach to this would be something like [Vault](https://www.vaultproject.io). This way you could pass secrets into `user-data` using [response wrapping](https://www.vaultproject.io/docs/concepts/response-wrapping). This would however require some commands to be run in order to unwrap the secrets. I'm not sure if this can currently be done with the default settings (maybe the jinja template make this possible, I don't know). For now I would be ok with keeping custom scripts disabled by default. Users can still enable the `custom_scripts` module by providing an appropriate `#cloud-config` document. If I come across a use case where using custom scripts is necessary I will open another issue/PR.

> if you agree with the rest, can you update the PR to include only write_files so we can merge it?

Yes I agree and I will amend the PR accordingly. I think `write_files` is very useful for writing credential files (as you mentioned) and I also think that it provides a useful way to use configuration files instead of commands for those that are currently using configuration files instead of commands.

I have added an additional annotation to the code.

File: config/cloud.cfg.d/10_vyos.cfg [hunk 1]
```diff
@@ -20,11 +20,11 @@ cloud_init_modules:
 # The modules that run in the 'config' stage
 cloud_config_modules:
   - vyos
+  - write_files
   - vyos_userdata
 
 # The modules that run in the 'final' stage
-cloud_final_modules:
-  - runcmd
```
Comment (codello) [inline config/cloud.cfg.d/10_vyos.cfg:27]: I also removed the `runcmd` module. The module only writes a script file that will not be executed unless the `scripts_user` module is also enabled.

**Comment by zdc** [comment]
Thanks!
I think that third-party software like the vault is a bit out of scope for Cloud-init, at least for now. Most likely, if someone will have control over a meta-data for deployment, a security breach will be huge enough to pass any protections.

For the future PRs, please see the details on how to contribute: https://docs.vyos.io/en/latest/contributing/development.html#submit-a-patch

Thanks! Merged.

**Pull request merged**